internal/controller/resourceset_manager.go

```diff
@@ -29,7 +43,66 @@ func (r *ResourceSetReconciler) SetupWithManager(mgr ctrl.Manager, opts Resource
 					predicate.AnnotationChangedPredicate{},
 				),
 			)).
+		Watches(
+			&fluxcdv1.ResourceSetInputProvider{},
+			handler.EnqueueRequestsFromMapFunc(r.requestsForChangeOf(inputsProviderIndexKey)),
+			builder.WithPredicates(exportedInputsChangePredicate),
+		).
 		WithOptions(controller.Options{
 			RateLimiter: opts.RateLimiter,
 		}).Complete(r)
 }
+
+func (r *ResourceSetReconciler) requestsForChangeOf(indexKey string) handler.MapFunc {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		log := ctrl.LoggerFrom(ctx)
+
+		var list fluxcdv1.ResourceSetList
+		if err := r.List(ctx, &list, client.MatchingFields{
+			indexKey: client.ObjectKeyFromObject(obj).String(),
+		}); err != nil {
+			log.Error(err, "failed to list objects for provider change")
+			return nil
+		}
+
+		reqs := make([]reconcile.Request, len(list.Items))
+		for i, rset := range list.Items {
+			reqs[i].NamespacedName = types.NamespacedName{Name: rset.Name, Namespace: rset.Namespace}
+		}
+
+		return reqs
+	}
+}
+
+func (r *ResourceSetReconciler) indexBy(kind string) func(o client.Object) []string {
+	return func(o client.Object) []string {
+		rs, ok := o.(*fluxcdv1.ResourceSet)
+		if !ok {
+			return nil
+		}
+
+		if len(rs.Spec.InputsFrom) == 0 {
+			return nil
+		}
+
+		results := make([]string, 0)
+		for _, k := range rs.Spec.InputsFrom {
+			if k.Kind == kind {
+				ns := rs.GetNamespace()
+				results = append(results, fmt.Sprintf("%s/%s", ns, k.Name))
+			}
+		}
+
+		return results
+	}
+}
+
+var exportedInputsChangePredicate = predicate.Funcs{
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		oldObj := e.ObjectOld.(*fluxcdv1.ResourceSetInputProvider)
+		newObj := e.ObjectNew.(*fluxcdv1.ResourceSetInputProvider)
```

Comment:
I'm not quite sure but it seems to be a pattern checking if `e.ObjectOld` or `e.ObjectNew` are `nil` and returning `false` in that case. I'm also an adept of not using defensive programming if you can avoid it, but I've seen this in many predicates of the Flux codebase. Why do we do that? Can these values actually be `nil` here? Would controller-runtime actually pass `nil` values here?

Reply:
Here we cast to a pointer while in Flux we cast to a value.

Reply:
I know we have those in Flux, but I don't see how an update event can send nil values, update means you always have old/new.